### PR TITLE
[main] moe(perf): Refactor CP layout organization for Qwen3.5-style hybrid attention model

### PR DESCRIPTION
**Pull request by yuzhongw-nvidia**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

PR for dev: https://github.com/NVIDIA/Megatron-LM/pull/6387

This PR adds fixed context-parallel (CP) layout support for long-context hybrid models that want to run most layers in `cp_partition_mode="contiguous"` while still keeping regular attention/GQA in the `zigzag` layout it requires.

image

The main goals are:

- Move CP layout conversion to the GQA/regular-attention boundary for Qwen3.5-style schedules. In the target GQA:GDN = 1:3 pattern, GDN chunkwise CP can consume contiguous layout directly, so only the less frequent GQA layers need to convert to zigzag and back. This reduces layout conversion frequency by roughly two thirds.
- Prebuild THD CP layout routes once per microbatch. THD layout conversion can then reuse precomputed split/routing metadata instead of rebuilding the route on the conversion hot path.

## Motivation

The previous fixed CP layout path effectively kept the model in zigzag layout because regular attention requires zigzag. That is correct for attention, but it is not optimal for Qwen3.5-style hybrid stacks where most layers are GDN layers.

For chunkwise CP, GDN can run in contiguous layout. This PR lets the model keep the broader stack in contiguous layout, then converts only when entering regular attention/GQA and converts the attention output back afterward. With a GQA:GDN = 1:3 schedule, this shifts layout shuffles away from the dense GDN path and places them on the sparser attention boundary.

For THD packed sequences, the route for a given microbatch is deterministic. This PR moves route construction into batch preparation and reuses the result for later `zigzag <-> contiguous` conversions.

### 1. CP layout conversion package

- Splits `megatron/core/context_parallel_layout.py` into a package:
  - `conversion.py`: SBHD/THD tensor conversion and `CpPartitionModeConverter`.
  - `routes.py`: THD route construction, encoding/decoding, lookup, and prebuild.
  - `metadata.py`: helpers for packed sequence layout metadata.
  - `policy.py`: stage-entry layout policy helpers.
  - `__init__.py`: compatibility exports for existing imports.
- Adds `convert_module_input_tensors_cp_partition_mode()` for module-boundary conversion:
  - converts module inputs to the required concrete layout;
  - returns a converter for mapping module outputs back to the caller layout;
  - rejects unsupported dense attention masks/bias and cross-attention key/value states.

### 2. Attention/GDN layout boundary

- Regular attention converts its input to zigzag when the caller layout is different, then converts the output back to the caller layout.
- GDN chunkwise CP can run directly in contiguous layout.
- GDN headwise CP keeps the existing zigzag requirement.
- This makes `cp_partition_mode="contiguous"` useful for Qwen3.5-style schedules: most GDN layers avoid conversion, while GQA/regular-attention layers handle their own layout boundary.

### 3. THD route prebuild

- Adds `prebuild_thd_cp_partition_routes()` to compute both THD conversion directions during batch construction.
- Updates GPT and Hybrid `get_batch` paths to annotate the stage-entry layout and prebuild THD routes before model forward.
- Updates route lookup so conversion can consume the prebuilt route. Missing routes still fall back to lazy construction with a warning for compatibility, but the intended path is prebuild during batch preparation.

### 4. Checks and unsupported cases

This PR adds two layers of validation.

First, `TransformerConfig` rejects unsupported configuration combinations early:

- `contiguous` CP currently requires `experimental_attention_variant` to be `gated_delta_net` or `dsv4_hybrid`.
- `dsv4_hybrid` with CP requires `contiguous`.
- `gated_delta_net` with `linear_cp_mode="headwise"` cannot use `contiguous`.
- `multi_latent_attention` outside `dsv4_hybrid` cannot use `contiguous`.

Second, runtime layout is carried through `PackedSeqParams.cp_partition_mode`, and attention variants validate the layout they receive:

- Regular attention requires zigzag.
- GDN chunkwise CP requires contiguous.
- GDN headwise CP requires zigzag.
- MTP validates its rolling path and rejects non-packed/SBHD contiguous layout until that path is implemented.

These checks are meant to fail early when tensor order would otherwise silently diverge from the module's layout assumption.

## Limitations / Future Work

- Dense attention masks, attention bias, and cross-attention key/value states are still rejected for CP layout conversion.
- Non-packed/SBHD contiguous MTP is intentionally not supported yet.
- Full-iteration CUDA graph is not supported for THD CP layout conversion today because torch-native NCCL A2A only supports fixed splits. Supporting this path will require GIN, and we plan to enable it through DeepEP / HybridEP dispatch.
- TP x CP conversion can be further optimized to use a single A2A communication. The current implementation still needs two communication steps.
- Auto conversion follow-up: <https://github.com/NVIDIA/Megatron-LM/pull/5664|[dev] moe(perf): Refactor CP layout - auto layout conversion by yuzhongw-nvidia · Pull Request #5664 · NVIDIA/Megatron-LM>

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/context_parallel_layout/conversion.py [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+
+    This helper performs the common "entry conversion" pattern used by modules
+    that need to consume a different CP layout than their caller supplied.  It
+    returns a converter for the opposite edge so the module output can be
+    converted back to the original input layout.
+    """
+    if cp_group is None or cp_group.size() <= 1:
+        return hidden_states, None
+
+    source_partition_mode = getattr(config, "cp_partition_mode", None)
```
Comment (Mellonta) [inline megatron/core/context_parallel_layout/conversion.py]: Is is intended that we are reading mode from config here?

File: megatron/core/transformer/transformer_config.py [hunk 2]
```diff
@@ -1321,6 +1328,15 @@ def __post_init__(self):
...
 
+        if self.cp_partition_mode not in ("zigzag", "contiguous"):
+            raise ValueError(f"Unsupported cp_partition_mode: {self.cp_partition_mode}")
+
+        if self.cp_partition_mode == "contiguous" and self.context_parallel_size > 1:
+            raise ValueError(
+                "cp_partition_mode='contiguous' with context parallelism requires a "
+                "contiguous-layout batch partitioning path, which is not currently available."
+            )
+
```
Comment (Mellonta) [inline megatron/core/transformer/transformer_config.py]: Why is it rejected?

File: megatron/core/transformer/transformer_config.py [hunk 3]
```diff
@@ -301,6 +301,13 @@ class TransformerConfig(ModelParallelConfig):
     'gated_delta_net' is a deprecated alias of 'gdn': it is normalized to 'gdn' in
     __post_init__ and emits a DeprecationWarning."""
 
+    cp_partition_mode: Literal["zigzag", "contiguous"] = "zigzag"
+    """How THD sequence rows are partitioned across context-parallel ranks.
+
+    Contiguous partitioning is defined only for THD inputs; BSHD context parallelism uses the
+    standard zigzag layout.
```
Comment (Mellonta) [inline megatron/core/transformer/transformer_config.py]: why is BSHD not supported?

**Review by Mellonta — COMMENTED** [review]
I didn't see how to configure linear attention to use the contiguous CP layout, is it not supported in this PR? Should we add separate CP partition-mode configs for linear and softmax attention and wire each module to its corresponding setting?

**Comment by Mellonta** [comment]
I think we also need a `_get_batch_on_this_cp_rank_contiguous`

File: megatron/core/context_parallel_layout/conversion.py (hunk 1 above)
Comment (yuzhongw-nvidia) [reply megatron/core/context_parallel_layout/conversion.py]: Yes, because we suppose the input is in global layout, i.e., `config.cp_partition_mode`.

File: megatron/core/context_parallel_layout/utils.py [hunk 4]
```diff
@@ -25,3 +25,21 @@ def get_packed_seq_params_cp_partition_cu_seqlens(
...
+def finalize_packed_seq_params(
+    packed_seq_params: Optional["PackedSeqParams"],
+) -> Optional["PackedSeqParams"]:
+    """Resolve CP metadata and prebuild the THD layout route for a microbatch."""
+    if packed_seq_params is None:
+        return None
+
+    # Keep these imports local: routes depends on this module for metadata access.
+    from megatron.core.context_parallel_layout.routes import prebuild_thd_cp_partition_routes
+    from megatron.core.packed_seq_params import resolve_cp_group
```
Comment (Mellonta) [inline megatron/core/context_parallel_layout/utils.py]: "resolve_cp_group" only available in dev?

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/3a1ab7f71380120824545abb5144623d0e495a6d

File: megatron/core/context_parallel_layout/utils.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    from megatron.core.parallel_state import get_context_parallel_group
+
+    cp_group = resolve_cp_group(get_context_parallel_group(), packed_seq_params)
+    packed_seq_params.cp_group = cp_group
```
Comment (Mellonta) [inline megatron/core/context_parallel_layout/utils.py]: Shall we remove this write back? it introduces some intrusive behaviors even when using default layout. e.g. --cp-comm-type a2a+p2p  and context_parallel_size == 1

File: megatron/core/packed_seq_params.py [hunk 6]
```diff
@@ -26,6 +34,8 @@ class PackedSeqParams:
     seq_idx: Tensor = None
     tokens_per_sample: int = None
     pad_between_seqs: bool = None
+    cp_partition_mode: Literal["zigzag", "contiguous"] = "zigzag"
+    cp_partition_route: Optional["ThdCpRoute"] = None
```
Comment (Mellonta) [inline megatron/core/packed_seq_params.py]: Does the new field break cuda graph?

File: megatron/core/context_parallel_layout/utils.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    prebuild_thd_cp_partition_routes(packed_seq_params, cp_group)
```
Comment (yuzhongw-nvidia) [inline megatron/core/context_parallel_layout/utils.py]:
```suggestion
    prebuild_thd_cp_partition_routes(packed_seq_params, cp_group)
```

File: megatron/core/packed_seq_params.py (hunk 6 above)
Comment (yuzhongw-nvidia) [reply megatron/core/packed_seq_params.py]: `cp_partition_route` is optional and defaults to None; existing recipes keep the default zigzag layout, so no conversion is triggered and the route remains unset, preserving existing CUDA graph behavior.

A non-empty ThdCpRoute is not CUDA-graph compatible yet, regardless of whether it is stored in PackedSeqParams, because torch-native NCCL A2A captures fixed split sizes. Supporting dynamic THD routes requires GIN, which we will leave as future work.

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/34d35a5b57eb92512ae6ac9674c715654bf6d185

**Review by Mellonta — APPROVED** [review]

**Review by zhongbozhu — APPROVED** [review]

**Comment by yuzhongw-nvidia** [comment]
Closed by mistake. Re-open in https://github.com/NVIDIA/Megatron-LM/pull/7128
